### ISSUE TITLE
`def settings` is in a private Eclipse object on master, how to access?

### ISSUE THREAD

**Issue by godenji** [typesafehub/sbteclipse#279]
Must be missing something: https://github.com/typesafehub/sbteclipse/blob/master/src/main/scala/com/typesafe/sbteclipse/core/Eclipse.scala

What is the point? Looks the above will compile and that's about it, every top level object/class is private o_O

`EclipseTransformerFactory` rewrites all break in attempted move to 4.0 RC2. Thanks for clues.

**Comment by benmccann** [comment]
I'm not sure I understand the problem that you're having. Can you post the code for what you're trying to do, what you expect to happen, and what actually happens?

**Comment by godenji** [comment]
@benmccann hey, the problem is that previously, on <= 3.0.0, `def setting(...)` was publicly available in [core.package.scala](https://github.com/typesafehub/sbteclipse/blob/v3.0.0/sbteclipse-core/src/main/scala/com/typesafe/sbteclipse/core/package.scala#L63) but is now in [core.Eclipse](https://github.com/typesafehub/sbteclipse/blob/master/src/main/scala/com/typesafe/sbteclipse/core/Eclipse.scala#L681) which is a private object.

Again, I must be missing something, but AFAICT core.Eclipse.scala is there to generate heat. What do you guys do with a source file where all top level class/object are private?

Basically I just need to access at least `def setting(...)` again as my rewrites fail without it:

    new EclipseTransformerFactory[RewriteRule] {
      override def createTransformer(ref: ProjectRef, state: State): Validation[RewriteRule] = {
        setting(crossTarget in ref, state).map{ct=>
          ....
        }
      }
    }

**Comment by benmccann** [comment]
What do you mean by "generate heat"? Are you asking where the class is used? [Here's one place](https://github.com/typesafehub/sbteclipse/blob/master/src/main/scala/com/typesafe/sbteclipse/core/EclipsePlugin.scala#L47)

You're welcome to submit a PR to make that class public if it helps you.

**Comment by godenji** [comment]
I mean that I didn't think one could access an object outside of the source file in which it is defined if marked as private -- therefore, generate heat in the Haskell sense of doing nothing ;-) Apparently not the case, must be package private by default.

Can issue a pull request, but presumably there's a reason it's private (as in there are things that need to be hidden from the user). If not, and it should be public, let me know, I'll do the necessary.

**Comment by benmccann** [comment]
You could also move the setting method back to package.scala. Or you could simply define your own in your code since it's only a few lines to copy

**Comment by godenji** [comment]
Those options work. I know you guys are trying to get 4.0 released, but it would be nice to do the "right" thing in terms of exposing (or not) the contents of the `Eclipse` object.

Maybe @jsuereth or @eed3si9n will have some feedback here re: what course to take if any. Not in a huge rush on my end, was just hoping to try the 4.0 RC with recently released Scala IDE 4.1

**Comment by benmccann** [comment]
I just simplified the method, so it's only a few lines if you want to copy it into your own project:

    def setting[A](key: SettingKey[A], state: State): A = key.get(structure(state).data).getOrElse {
      throw new IllegalStateException("Undefined setting '%s in %s'!".format(key.key, key.scope))
     }

**Comment by godenji** [comment]
@benmccann thanks, I'll give the RC a spin again ;-)
